### ISSUE TITLE
Feature Request: input/output - Imagick objects, binary blobs, GD objects?

### ISSUE THREAD

**Issue by Kadigan** [rosell-dk/webp-convert#235]
Would it be possible to have your converter handle these things? In order of importance to me, it'd be Imagick, then blobs (binary file contents), then GD objects.

The way I'd expect this feature to work would be that instead of providing a $src and $dst, I could simply provide one param - an object of either Imagick or GD, or a binary blob - and have the process return the result in kind?

Being able to handle Imagick objects is important to me - I need to perform colorspace conversions, and I want to do it using Imagick even if it doesn't support WebP.

Being able to handle binary blobs is important to me - I sometimes need to process images in environments where saving files to disk is impossible/prohibited, and I would much rather receive a binary blob in response.

The inclusion of GD objects - simply for completeness' sake - not important to me personally, since the GD library doesn't do well w/ non-sRGB images, so I don't use it much.

I am aware that using any of these would probably make the process incompatible w/ some of the methods (notably anything exec-related).

**Comment by Kadigan** [comment]
Alternatively, it might be helpful if your converter first brought images to sRGB. As far as I'm aware, PHP's GD library doesn't support color profiling, and if you want Webp support w/ ICC in Imagick, then you need to build libwebp specifically with lcms -- so this is something that might trip you up, if you're not aware of the issues.

**Comment by rosell-dk** [comment]
Hi Kadigan,

Thanks for your suggestions.

I'm not inclined to provide multiple signatures for the `::convert()` method. But I could simply provide a new `::convertBlob()` method for receiving and returning blobs. This goes better with doc comments. The strongly typing via doc comments are useful because they help track errors (by the help of PHPStan and [Scrutinizer](https://scrutinizer-ci.com/g/rosell-dk/webp-convert/)).

Similar, we could have a `::convertObject()` method on the Imagick class, which receives and returns an Imagick object.

I'm wondering if there would be need for a an output that is different than the input. This could be implemented with a `return-type` option. However, this would require loosing up the return type in the doc-comments, so there is a price. I'm guessing there is no real need as converting a blob or an image to a file is trivial, as you say.

**Comment by rosell-dk** [comment]
Yes, it seems blobs would be a problem for the exec converters. We could of course save the blob as a file first, do the conversion and return the content of the file. But this will not work in environments where storing files is disallowed. So this apparently won't do, as this is exactly why we would want a `::convertBlob()` method. But on the other hand, it *will* do if there is just a single place we can write - for example /tmp.

It is possible to pass a blob to ImageMagick via the command line [1](https://www.imagemagick.org/discourse-server/viewtopic.php?t=26918) [2](https://www.imagemagick.org/Usage/files/#inline). However, this will only work for [small files](https://serverfault.com/a/163390). My system allows ~2MB on the command line.

Perhaps some command line wizardry could be used to overcome this. The main problem seems to be finding a place to store the blob that will be accessible with a command in the shell when a regular file isn't an option.

**Comment by Kadigan** [comment]
No issue from me on having different methods for these things, so long as they're nicely summed up somewhere in the main readme. All I ask is for the functionality to be there.

Being able to process images w/o having access to writeable disk space is an uncommon requirement - true... but when you find you have a problem like that, there aren't that many solutions for it.

As for processing blobs - thankfully, PHP has its own implementation of ImageMagick (the `Imagick` class), and it can both read and output binary blobs just fine (there are built-in methods for doing both). GD2 can also process blobs directly - it has `imagecreatefromstring()` if memory serves (makes you wonder why other format creators are exposed, if you can read every supported image time with a `imagecreatefromstring( file_get_contents( ... ) )`), and you can catch output w/ output buffering.

**Comment by rosell-dk** [comment]
Hm. Looking more closely in the source, I realize that such change will be api breaking. The `AbstractConverter` class, which the rest of the converter classes are based upon takes `$source` and `$destination` in the constructor. Or, alternatively, I must do as you suggest and fake multiple signatures for that constructor.

**Comment by rosell-dk** [comment]
Forget that last post. I can simply set an (internal) option for ignoring those arguments

**Comment by rosell-dk** [comment]
I dropped this ball. Is it still relevant?

**Comment by Kadigan** [comment]
To be honest I haven't done a lot of work with this - I entirely dropped the work due to having color profile issues, and to a lesser extent because the produced .webp files weren't offering great savings. The functionality I asked for was more of a quality-of-life thing, and I'd still like to see it included at some point... but like I said, due to various issues, I dropped WebP support entirely -- I couldn't get the colours right, and when I eventually did, the savings were marginal (to the tune of maybe 10% difference).
